torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -488,6 +488,20 @@ def inc_and_set(fut):
     return ret_future
 
 
+@rpc.functions.async_execution
+def async_cuda_sleep_and_set_to_one(t):
+    device = t.device
+    original_stream = torch.cuda.current_stream(device)
+    new_stream = torch.cuda.Stream(device)
+    new_stream.wait_stream(original_stream)
+    with torch.cuda.stream(new_stream):
+        torch.cuda._sleep(int(1000 * get_cycles_per_ms()))
+        t.fill_(1)
+        fut = Future(devices=[device])
+        fut.set_result(t)
```

Comment:
do we also want to test the case where there is a real callback, as that's the most common case I guess?

Reply:
I think we should test one thing at a time. Using a "real callback" is a CPU-only aspect and I assume you added a test for it back in the days. Here we're interested about CUDA only, thus as long as we correctly control the streams it doesn't matter which thread is doing it.

Reply:
hmm, I didn't follow the argument here. Looks like we agree that, in normal use cases, there can be callbacks on the CUDAFuture, which might run in a different thread, but we decided to intentionally not testing that? And we don't test because we believe CUDA streams should be correctly managed even if we switch thread? If the goal is to ensure async execution works with CUDAFuture, shouldn't we test all common usage of async executions? The one being tested here is indeed one common usage. Another one can be using a async_execution to wait for nested RPC.

```python
@rpc.functions.async_execution
def test_async_execution_cuda_callback():
        def cb():
            # do sth

        return rpc.async(...).then(cb)
```

It looks good to me if we plan to add that test in a followup PR.

Reply:
What I was trying to say is that if we make our unit tests cover multiple aspects all at once they're in fact becoming integration tests (and are heavier and noisier). Specifically here I was decomposing the problem in two parts: we should both tests the various ways in which an async remote function could "produce" a future, and then test that we correctly propagate the CUDA information contained in that future. The first part isn't new, so I assume we are already testing it somewhere (are we?), hence I focused only on the second part.

Though if you think it's important I'll add a test like the one you suggested.

Reply:
> The first part isn't new, so I assume we are already testing it somewhere (are we?), hence I focused only on the second part.

I recall I did add tests for nested RPC, but not in the context for CUDA RPC + `async_execution`. I'd assume we didn't have that test, because before your fix `async_execution` does not work with CUDA RPC yet.